### PR TITLE
Idea deps optimization

### PR DESCRIPTION
**Pull request by nskvortsov**
This pull-request is a re-implementation of https://github.com/gradle/gradle/pull/3321

Reduce number of dependencies in generated or imported Idea projects by using the "scope inheritance" similar to what Gradle have. E.g., compile and provided scope dependencies are always included in test compile and runtime, etc.

This reduces number of dependencies twice on average, speeding up project import, sync and improving overall IntelliJ performance on middle to large projects.

### CODE REVIEW

File: subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/model/internal/IdeaDependenciesProvider.java [hunk 1]
```diff
@@ -78,34 +73,128 @@ public IdeaDependenciesProvider(ServiceRegistry serviceRegistry) {
...
+    private Collection<Dependency> optimizeDeps(Collection<ModuleDependency> deps) {
+        Set<Dependency> result = Sets.newLinkedHashSet();
+        Multimap<String, GeneratedIdeaScope> namesToScopes = MultimapBuilder.hashKeys().linkedHashSetValues().build();
+        for (ModuleDependency dep : deps) {
+            namesToScopes.put(dep.getName(), GeneratedIdeaScope.valueOf(dep.getScope()));
+        }
+
+        for (ModuleDependency dep : deps) {
+            String moduleName = dep.getName();
+            Collection<GeneratedIdeaScope> ideaScopes = namesToScopes.get(moduleName);
```
Comment (oehme) [inline subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/model/internal/IdeaDependenciesProvider.java]: ❌ There's lots of duplicated code between this method and the next.

Comment (nskvortsov) [reply subprojects/ide/src/main/java/org/gradle/plugins/ide/idea/model/internal/IdeaDependenciesProvider.java]: sorry for that. I've added the patch to deduplicate the code.

**Comment by nskvortsov** [comment]
Is there anything else I can improve?

**Comment by oehme** [comment]
Thank you, this looks good from my side, I'll merge it later this week.

**Comment by oehme** [comment]
I rebased the changes on master and did a bit of cleanup. I'll merge this once CI is green. If you have time I'd appreciate if you could give the final changes a look.

**Comment by oehme** [comment]
Thanks a lot Nikita!

**Pull request merged**